File: ds_4_linkedlist/linkedlist_3_two_points/L19-m.cpp
//
// Created by LiHeng on 2019/7/28.
//

//双指针技术在链表中的应用
//leetcode  19 删除倒数第n个节点
#include <iostream>
#include <cassert>

using  namespace std ;

struct ListNode{
    int val ;
    ListNode* next ;
    ListNode(int x ): val(x) , next(NULL) {}
};

//设置虚拟头节点
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        //一直约定的是back是右边的，pre是靠左边的
        ListNode* pre = dummy;
        ListNode* back = dummy;
        for(int i=0; i<n; ++i)
            back = back->next;
        while(back->next != NULL){
            pre = pre->next;
            back = back->next;
        }
        pre->next = pre->next->next;
        return dummy->next;
    }
};

int main(){

    return  0 ;
}

//练习 leetcode 61 143 234

File: python-algorithm/ds_4_linkedlist/linkedlist_3_two_points/L19-m.py
#--coding:utf-8--
'''
@Time   : 2020/8/14
@Author : Heng Li
@Email  : liheng_lulia@163.com
'''

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        dummy = ListNode(-1)
        dummy.next = head
        slow = dummy
        fast = dummy
        for i in range(n+1):
            fast = fast.next
        while fast:
            slow = slow.next
            fast = fast.next
        slow.next = slow.next.next
        return dummy.next